### PR TITLE
[MISC] IPC Coupling Unit Tests

### PR DESCRIPTION
**Pull request by Roushelfy**
## Description
This PR strengthens IPC coupling regression coverage by adding strict white-box + black-box checks across cloth, rigid articulation, FEM grasp/retrieve, and momentum conservation scenarios.

### 1) `test_ipc_cloth` (`n_envs=0,2`)
- Scenario:
  - Cloth + rigid box + FEM sphere under gravity with IPC enabled.
- Checks:
  - Cloth entity is uniquely registered in IPC metadata (per env).
  - Cloth vertex kinematics follow expected gravity-driven position/velocity updates.
- Expected:
  - Unique IPC registration passes.
  - Position/velocity errors remain within tolerance.

### 2) `test_ipc_two_way_revolute` (coupling type × `fixed_base=True/False`)
- Scenario:
  - `two_cube_revolute.urdf`, setup aligned with `examples/IPC_Solver/ipc_test_joint.py`:
    - `dt=1e-2`, gravity, plane, rigid collision disabled, aligned IPC solver options.
- Checks:
  - Moving link is actually added to IPC metadata and mapped in `_link_to_abd_slot`.
  - `external_articulation`: IPC transform stays close to Genesis transform.
  - `two_way_soft_constraint`: commanded joint trajectory is tracked (correlation + motion amplitude).
  - Free-base branch:
    - `external_articulation`: base drops in `z`.
    - `two_way_soft_constraint`: base position changes (instead of strict drop).
- Expected:
  - All registration and mode-specific behavior checks pass for all parameter combinations.

### 3) `test_ipc_two_way_prismatic` (coupling type × `fixed_base=True/False`)
- Scenario:
  - `two_cube_prismatic.urdf`, same aligned setup as revolute.
- Checks:
  - Same structure as revolute:
    - IPC add/slot checks,
    - strict transform consistency for `external_articulation`,
    - target tracking (correlation + amplitude) for `two_way_soft_constraint`,
    - mode-specific free-base checks.
- Expected:
  - All combinations pass with expected coupling behavior.

### 4) `test_ipc_cloth_gravity_freefall`
- Scenario:
  - Cloth-only freefall under gravity.
- Checks:
  - Cloth is uniquely added to IPC.
  - Physical displacement matches gravity expectation.
  - IPC-to-Genesis retrieve consistency (centroid match).
- Expected:
  - Physics error within tolerance.
  - Retrieve mismatch below millimeter-level threshold.

### 5) `test_ipc_link_filter_strict`
- Scenario:
  - Revolute two-cube robot with `two_way_soft_constraint`, link filter set to only `moving`.
- Checks:
  - `_ipc_link_filters` content is exactly expected.
  - `moving` is in IPC; `base` is excluded.
  - `_link_to_abd_slot` matches filter outcome.
- Expected:
  - Only filtered links are coupled into IPC.

### 6) `test_ipc_ipc_only_cube_freefall_height_drop`
- Scenario:
  - Single rigid cube with `ipc_only`, initialized in air.
- Checks:
  - Cube base link exists in IPC metadata and `_link_to_abd_slot`.
  - Simulation result is written back to Genesis state.
- Expected:
  - Cube final height is significantly lower than initial height.

### 7) `test_ipc_robot_fem_grasp_retrieve_lift_strict` (2 coupling modes)
- Scenario:
  - Franka grasps and lifts FEM box (approach, descend, close fingers, lift).
- Checks:
  - FEM is added to IPC metadata.
  - Required finger links are present in IPC rigid metadata.
  - Mode-specific link behavior:
    - `two_way_soft_constraint`: strict filter match.
    - `external_articulation`: required links included.
  - Retrieve consistency between IPC FEM centroid and Genesis FEM centroid (before/after).
  - Final lift success.
- Expected:
  - Coupling + retrieve checks pass.
  - FEM final centroid `z` increases by at least `0.1m`.

### 8) `test_ipc_motion_final_relative_error_below_2pct`
- Scenario:
  - Zero-gravity rigid+FEM momentum exchange case.
- Checks:
  - Both rigid and FEM are correctly coupled into IPC.
  - Final-step total momentum relative error.
- Expected:
  - Final relative error `<= 2%`.

## Test Result
- `uv run pytest tests/test_ipc.py -n 0`
- **16 passed**

### CODE REVIEW

File: tests/test_ipc.py [hunk 1]
```diff
@@ -1,44 +1,157 @@
...
 
 import genesis as gs
 
 from .utils import get_hf_dataset
 
 try:
-    import uipc
+    from uipc import builtin
+    from uipc.backend import SceneVisitor
+    from uipc.geometry import SimplicialComplexSlot, apply_transform, merge
```
Comment (duburcqa) [inline tests/test_ipc.py]: This is wrong. Move these imports after this block, without changing this one.

File: tests/test_ipc.py [hunk 2]
```diff
@@ -1,44 +1,157 @@
...
+
+def _get_fem_solver_entity_idx(scene, fem_entity):
+    for idx, entity in enumerate(scene.sim.fem_solver._entities):
+        if entity is fem_entity:
+            return idx
+    raise AssertionError(f"FEM entity {fem_entity} not found in FEM solver entity list.")
+
+
+def _get_entity_base_z(entity):
+    pos = np.asarray(entity.get_pos().detach().cpu().numpy())
```
Comment (duburcqa) [inline tests/test_ipc.py]: Never do `.detach().cpu().numpy()`. Use `tensor_to_array`.

File: tests/test_ipc.py [hunk 3]
```diff
@@ -226,99 +351,118 @@ def test_ipc_two_way_revolute(n_envs, coupling_type, show_viewer):
...
+            final_base_z = _get_entity_base_z(robot)
+            assert final_base_z <= initial_base_z - 0.03, (
+                f"Free-base revolute robot base did not drop enough: dz={initial_base_z - final_base_z:.6f} m."
+            )
+        else:
+            final_base_pos = np.asarray(robot.get_pos().detach().cpu().numpy()).reshape(-1, 3)[0]
+            base_move = np.linalg.norm(final_base_pos - initial_base_pos)
+            assert base_move > 1e-3, (
+                f"Two-way free-base revolute base position did not change enough: dpos={base_move:.6f} m."
+            )
```
Comment (duburcqa) [inline tests/test_ipc.py]: Do not this. Use our own `assert_allclose`. And never print the values that are being compared. pytest will do it for you automatically (this is not the case here).

File: tests/test_ipc.py [hunk 4]
```diff
@@ -328,71 +472,48 @@ def test_ipc_two_way_prismatic(n_envs, coupling_type, show_viewer):
...
-                    print(f"Step {i}: Rotation difference (rad) = {rot_diff:.6f}")
-                    assert rot_diff < 0.1, f"Rotation difference too large: {rot_diff}"
+                    if coupling_type == "external_articulation":
+                        assert rot_diff < 0.1, f"Rotation difference too large: {rot_diff}"
+
+    if coupling_type == "two_way_soft_constraint":
+        q_hist = np.asarray(qpos_history)
+        tgt_hist = np.asarray(target_history)
+        corr = np.corrcoef(q_hist, tgt_hist)[0, 1]
+        assert corr > 0.4, f"Two-way prismatic target tracking correlation too low: corr={corr:.6f}."
```
Comment (duburcqa) [inline tests/test_ipc.py]: Similarly, never print yourself values involved in the right part of asserts.

File: tests/test_ipc.py
Comment (duburcqa) [inline tests/test_ipc.py:1]: Please start from the unit test in my PR and improve them. They have been updated to (mostly) follow our guidelines already.

https://github.com/duburcqa/Genesis/blob/9cbbe051b843e3ea675a6d0ccdf69d896f8ef78d/tests/test_ipc.py

File: tests/test_ipc.py [hunk 5]
```diff
@@ -440,37 +556,370 @@ def get_cloth_vertex_positions(scene):
...
+    total_p_history = []
+
+    test_time = 0.30
+    n_steps = int(test_time / dt)
+
+    for _ in range(n_steps):
+        rigid_pos = (
+            scene.sim.rigid_solver.get_links_pos(links_idx=rigid_link_idx, ref="link_com").detach().cpu().numpy()
+        )
+        rigid_pos = rigid_pos.flatten()[:3]
```
Comment (duburcqa) [inline tests/test_ipc.py]: Do not use flatten. Keep the batch dimension if possible. But if this is too complicated, explicitly do `[0]`

File: tests/test_ipc.py [hunk 6]
```diff
@@ -440,37 +556,370 @@ def get_cloth_vertex_positions(scene):
...
+
+        if fem_prev_pos is None:
+            fem_vertex_velocities = np.zeros_like(fem_vertex_positions)
+        else:
+            fem_vertex_velocities = (fem_vertex_positions - fem_prev_pos) / dt
+        fem_prev_pos = fem_vertex_positions.copy()
+
+        fem_linear_momentum = np.sum(fem_vertex_masses[:, np.newaxis] * fem_vertex_velocities, axis=0)
+        total_linear_momentum = rigid_linear_momentum + fem_linear_momentum
+        total_p_history.append(np.asarray(total_linear_momentum).flatten().copy())
```
Comment (duburcqa) [inline tests/test_ipc.py]: Remove `np.asarray` and `.copy()`, they are superfluous.

File: tests/test_ipc.py [hunk 7]
```diff
@@ -440,37 +556,370 @@ def get_cloth_vertex_positions(scene):
...
+    assert (0, rigid_link_idx) in scene.sim.coupler._link_to_abd_slot, "Missing ABD slot for rigid cube base link."
+
+    rigid_cube.set_dofs_velocity((4.0, 0.0, 0.0, 0.0, 0.0, 0.0))
+
+    blob_radius = blob.morph.radius
+    blob_rho = blob.material.rho
+    fem_total_mass = (4.0 / 3.0) * np.pi * (blob_radius**3) * blob_rho
+
+    mass_field = scene.sim.rigid_solver.links_info.inertial_mass
+    if hasattr(mass_field, "to_numpy"):
```
Comment (duburcqa) [inline tests/test_ipc.py]: This cannot be false. Anyway, use your helper for this: `qd_to_numpy`.

File: tests/test_ipc.py [hunk 8]
```diff
@@ -440,37 +556,370 @@ def get_cloth_vertex_positions(scene):
...
+    assert rigid_link_idx in ipc_links, "Rigid cube base link was not added to IPC."
+    assert (0, rigid_link_idx) in scene.sim.coupler._link_to_abd_slot, "Missing ABD slot for rigid cube base link."
+
+    rigid_cube.set_dofs_velocity((4.0, 0.0, 0.0, 0.0, 0.0, 0.0))
+
+    blob_radius = blob.morph.radius
+    blob_rho = blob.material.rho
+    fem_total_mass = (4.0 / 3.0) * np.pi * (blob_radius**3) * blob_rho
+
+    mass_field = scene.sim.rigid_solver.links_info.inertial_mass
```
Comment (duburcqa) [inline tests/test_ipc.py]: We have a public getter for this `get_links_inertial_mass`. What are you accessing internals directly? Never do this unless strictly necessary. This makes our unit tests harder to maintain and less useful for the interested users because they cannot replicated what is done her in their own script.

File: tests/test_ipc.py [hunk 9]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
+        x_next = _get_ipc_positions(scene, solver_type="cloth", idx=cloth_entity_idx, env_idx=0)
         assert x_next is not None
         x_next = x_next[0, 2]
 
-        # Expected displacement: dx = v_n * dt + 0.5 * g * dt^2
-        expected_dx = v_n * dt - g * dt * dt  # Negative because gravity is -g
+        expected_dx = v_n * dt - g * dt * dt
         expected_x_next = x_n + expected_dx
+        pos_error = abs(x_next - expected_x_next) / abs(expected_dx) if abs(expected_dx) > 1e-6 else 0.0
+        assert pos_error < tolerance
```
Comment (duburcqa) [inline tests/test_ipc.py:258]: assert_allclose

File: tests/test_ipc.py [hunk 10]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
 
-        # Validate position
-        pos_error = abs(x_next - expected_x_next) / abs(expected_dx) if abs(expected_dx) > 1e-6 else 0
-        assert pos_error < tolerance, f"Step {step}: Position error {pos_error * 100:.2f}% exceeds tolerance"
-
-        # Calculate velocity: v_{n+1} = (x_{n+1} - x_n) / dt
         v_next = (x_next - x_n) / dt
+        expected_v_next = v_n - g * dt
+        vel_error = abs(v_next - expected_v_next) / abs(expected_v_next) if abs(expected_v_next) > 1e-6 else 0.0
+        assert vel_error < tolerance
```
Comment (duburcqa) [inline tests/test_ipc.py:263]: assert_allclose

File: tests/test_ipc.py [hunk 11]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
+                            _rotmat_to_euler(ipc_transform[:3, :3]),
+                            atol=0.1,
+                        )
+
+    if coupling_type == "two_way_soft_constraint":
+        q_hist = np.asarray(qpos_history)
+        tgt_hist = np.asarray(target_history)
+        corr = np.corrcoef(q_hist, tgt_hist)[0, 1]
+        assert corr > 0.85
+        assert np.ptp(q_hist) > 0.7
```
Comment (duburcqa) [inline tests/test_ipc.py:369]: Numpy should accept passing `qpos_history`, `target_history` directly. Wrapping via `np.asarray` is irrelevant.

File: tests/test_ipc.py [hunk 12]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
+
+    if not fixed_base:
+        if coupling_type == "external_articulation":
+            final_base_z = _get_entity_base_z(robot)
+            assert final_base_z <= initial_base_z - 0.03
+        else:
+            final_base_pos = tensor_to_array(robot.get_pos()).reshape(-1, 3)[0]
```
Comment (duburcqa) [inline tests/test_ipc.py:376]: Avoid reshape.

File: tests/test_ipc.py [hunk 13]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
+                            _rotmat_to_euler(genesis_transform[:3, :3]),
+                            _rotmat_to_euler(ipc_transform[:3, :3]),
+                            atol=0.1,
+                        )
+
+    if coupling_type == "two_way_soft_constraint":
+        q_hist = np.asarray(qpos_history)
+        tgt_hist = np.asarray(target_history)
+        corr = np.corrcoef(q_hist, tgt_hist)[0, 1]
+        assert corr > 0.4
```
Comment (duburcqa) [inline tests/test_ipc.py:480]: same.

File: tests/test_ipc.py [hunk 14]
```diff
@@ -115,284 +233,271 @@ def get_cloth_vertex_positions(scene):
...
+        assert np.ptp(q_hist) > 0.12
+
+    if not fixed_base:
+        if coupling_type == "external_articulation":
+            final_base_z = _get_entity_base_z(robot)
+            assert final_base_z <= initial_base_z - 0.03
+        else:
+            final_base_pos = tensor_to_array(robot.get_pos()).reshape(-1, 3)[0]
```
Comment (duburcqa) [inline tests/test_ipc.py:488]: same.

File: tests/test_ipc.py [hunk 15]
```diff
@@ -440,37 +540,350 @@ def get_cloth_vertex_positions(scene):
...
+        total_linear_momentum = rigid_linear_momentum + fem_linear_momentum
+        total_p_history.append(total_linear_momentum)
+
+        scene.step()
+
+    total_p_history = np.array(total_p_history)
+    expected_momentum = np.array([4.0 * rigid_mass, 0.0, 0.0])
+    final_relative_error = np.linalg.norm(total_p_history[-1] - expected_momentum) / np.linalg.norm(expected_momentum)
+
+    assert final_relative_error <= 0.02
```
Comment (duburcqa) [inline tests/test_ipc.py:889]: assert_allclose

File: tests/test_ipc.py [hunk 16]
```diff
@@ -440,37 +540,350 @@ def get_cloth_vertex_positions(scene):
...
+        fem_prev_pos = fem_vertex_positions.copy()
+
+        fem_linear_momentum = np.sum(fem_vertex_masses[:, np.newaxis] * fem_vertex_velocities, axis=0)
+        total_linear_momentum = rigid_linear_momentum + fem_linear_momentum
+        total_p_history.append(total_linear_momentum)
+
+        scene.step()
+
+    total_p_history = np.array(total_p_history)
+    expected_momentum = np.array([4.0 * rigid_mass, 0.0, 0.0])
```
Comment (duburcqa) [inline tests/test_ipc.py:886]: Casting to numpy array is useless.

File: tests/test_ipc.py [hunk 17]
```diff
@@ -440,37 +540,350 @@ def get_cloth_vertex_positions(scene):
...
+    rigid_prev_pos = None
+    total_p_history = []
+
+    test_time = 0.30
+    n_steps = int(test_time / dt)
+
+    for _ in range(n_steps):
+        rigid_pos = tensor_to_array(
+            scene.sim.rigid_solver.get_links_pos(links_idx=rigid_link_idx, ref="link_com")
+        ).reshape(-1, 3)[0]
```
Comment (duburcqa) [inline tests/test_ipc.py:851]: Avoid reshape.

File: tests/test_ipc.py [hunk 18]
```diff
@@ -440,37 +540,350 @@ def get_cloth_vertex_positions(scene):
...
+        if rigid_prev_pos is None:
+            rigid_vel = np.array([4.0, 0.0, 0.0])
```
Comment (duburcqa) [inline tests/test_ipc.py:853]: What is this hard-coded value?

File: tests/test_ipc.py [hunk 19]
```diff
@@ -440,37 +540,350 @@ def get_cloth_vertex_positions(scene):
...
+        else:
+            rigid_vel = (rigid_pos - rigid_prev_pos) / dt
+        rigid_prev_pos = rigid_pos.copy()
+        rigid_linear_momentum = rigid_mass * rigid_vel
+
+        fem_proc_geo = _get_ipc_processed_geometry(scene, solver_type="fem", idx=fem_entity_idx, env_idx=0)
+        assert fem_proc_geo is not None
+        fem_vertex_positions = fem_proc_geo.positions().view().reshape(-1, 3)
```
Comment (duburcqa) [inline tests/test_ipc.py:861]: You already added a helper for this.

File: tests/test_ipc.py [hunk 20]
```diff
@@ -1,44 +1,162 @@
...
 try:
     import uipc
 except ImportError:
     pytest.skip("IPC Coupler is not supported because 'uipc' module is not available.", allow_module_level=True)
 
+from uipc import builtin
+from uipc.backend import SceneVisitor
+from uipc.geometry import SimplicialComplexSlot, apply_transform, merge
+
+_ = uipc
```
Comment (duburcqa) [inline tests/test_ipc.py:18]: What is this? Remove it.

File: tests/test_ipc.py [hunk 21]
```diff
@@ -1,44 +1,162 @@
...
+
+def _iter_ipc_simplicial_geometries(scene):
+    visitor = SceneVisitor(scene.sim.coupler._ipc_scene)
+    for geo_slot in visitor.geometries():
+        if isinstance(geo_slot, SimplicialComplexSlot):
+            yield geo_slot.geometry()
+
+
+def _get_processed_geo(geo):
+    proc_geo = geo
```
Comment (duburcqa) [inline tests/test_ipc.py:29]: Avoid assignment that just rename variable.

File: tests/test_ipc.py [hunk 22]
```diff
@@ -1,44 +1,162 @@
...
+        elif solver_type in ("fem", "cloth"):
+            idx_attr = meta_attrs.find("entity_idx")
+        else:
+            return None
+
+        if not idx_attr:
+            return None
+        idx = int(str(idx_attr.view()[0]))
+        return (solver_type, env_idx, idx)
+    except Exception:
```
Comment (duburcqa) [inline tests/test_ipc.py:82]: Do not catch any exception here.

File: tests/test_ipc.py [hunk 23]
```diff
@@ -1,44 +1,162 @@
...
+            idx_attr = meta_attrs.find("link_idx")
+        elif solver_type in ("fem", "cloth"):
+            idx_attr = meta_attrs.find("entity_idx")
+        else:
+            return None
+
+        if not idx_attr:
+            return None
+        idx = int(str(idx_attr.view()[0]))
+        return (solver_type, env_idx, idx)
```
Comment (duburcqa) [inline tests/test_ipc.py:81]: Why is there so many edge-cases? There is something wrong here.

File: tests/test_ipc.py [hunk 24]
```diff
@@ -1,44 +1,162 @@
...
+        if entity is fem_entity:
+            return idx
+    raise AssertionError
+
+
+def _get_entity_base_z(entity):
+    pos = tensor_to_array(entity.get_pos())
+    if pos.ndim == 1:
+        return float(pos[2])
+    return float(pos.reshape(-1, 3)[0, 2])
```
Comment (duburcqa) [inline tests/test_ipc.py:139]: WTF is happening here.

File: tests/test_ipc.py [hunk 25]
```diff
@@ -1,44 +1,162 @@
...
+    singular = sy < 1e-6
+    if not singular:
+        x = np.arctan2(rot_mat[2, 1], rot_mat[2, 2])
+        y = np.arctan2(-rot_mat[2, 0], sy)
+        z = np.arctan2(rot_mat[1, 0], rot_mat[0, 0])
+    else:
+        x = np.arctan2(-rot_mat[1, 2], rot_mat[1, 1])
+        y = np.arctan2(-rot_mat[2, 0], sy)
+        z = 0.0
+    return np.array([x, y, z])
```
Comment (duburcqa) [inline tests/test_ipc.py:153]: Don't do this. We have a helper for this.
